Add Uniform State Preparation

Hirmay opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

Closes #11735

### Details and comments

Hirmay committed:
Update state_preparation.py

Hirmay commented:
Just an update I have created a draft pull request here for the original issue #11735

Hirmay commented:
@woodsp-ibm I have put this as draft as you mentioned before. But I don't understand the errors in some of these checks. Can you please give some insight regarding this?

woodsp-ibm commented:
One aspect is the formatting of the code, the style is checked via `black` you need to run that on your code locally - see https://github.com/Qiskit/qiskit/blob/main/CONTRIBUTING.md#style-and-lint

> lint: commands[0]> black --check qiskit test tools examples setup.py
> would reformat /home/vsts/work/1/s/qiskit/circuit/library/data_preparation/state_preparation.py
> Oh no! 💥 💔 💥
> 1 file would be reformatted, 1257 files would be left unchanged.

The other is a cyclic import - I copied some of the error below. You have add this import `from qiskit.visualization import circuit_drawer` but I did not see anything in the added code using it and it would seem strange to import visualization logic here anyway. I would remove that and it most likely will eliminate the cyclic error in doing so.

```
  File "/home/vsts/work/1/s/qiskit/circuit/library/data_preparation/state_preparation.py", line 31, in <module>
    from qiskit.visualization import circuit_drawer
  File "/home/vsts/work/1/s/qiskit/visualization/__init__.py", line 259, in <module>
    from .circuit import circuit_drawer
  File "/home/vsts/work/1/s/qiskit/visualization/circuit/__init__.py", line 15, in <module>
    from .circuit_visualization import circuit_drawer
  File "/home/vsts/work/1/s/qiskit/visualization/circuit/circuit_visualization.py", line 39, in <module>
    from qiskit.circuit import ControlFlowOp, Measure
ImportError: cannot import name 'ControlFlowOp' from partially initialized module 'qiskit.circuit' (most likely due to a circular import) (/home/vsts/work/1/s/qiskit/circuit/__init__.py)

================================================================================
The above traceback was encountered during test discovery which imports all the found test modules in the specified test_path.
```

You will need to add unit test(s) too, its good to have already as it will show the code is working as intended.

Hirmay committed:
Update state_preparation.py

Hirmay committed:
Merge branch 'main' into patch-2

Hirmay committed:
Add files via upload

Hirmay committed:
Update __init__.py

Hirmay commented:
Hi @woodsp-ibm, I have done the changes as you suggested and even tested it locally. Please have a look into this and suggest future steps. 

Thank you!

Edit: Sorry Steve, the tests passed locally and I tried fixing some of these ones, but I am unable to understand the Coverage report check which is causing the errors in later tests.

Hirmay committed:
Update test_generalized_uniform_superposition_gate.py

Hirmay committed:
Update test_generalized_uniform_superposition_gate.py

woodsp-ibm commented:
It seems a test, that is a general test for equivalence, is picking up on your new "Gate" and failing. If you look that test has a exclusion list https://github.com/Qiskit/qiskit/blob/a7cc16624a719332b8ad5ab764425dca33f3952c/test/python/circuit/test_gate_definitions.py#L259 that already excludes StatePreparation - maybe this new one needs to be added there.

Since its failing on num_qubits, I was looking at StatePreparation and how it handles it there. Could this new one be made more compatible with that - ie compute for M the smallest n (where I think n there is synonomous with num_qubits right?) and let one expand things out otherwise by specifying num_qubits explicitly. I get that one can require num_qubits to always be specified - just wondering if it was done more like StatePreparation whether that would be good.

Hirmay commented:
Thank you, Steve, for your helpful response. So, if the `StatePreparation` is being excluded, then regardless of how they handle it, it won't matter, correct? Because there will be no `TestGateEquivalenceEqual` test case performed on that. Let me know if I'm not understanding it properly or missing something. Then should I just try running the tests with this gate also included and see what happens?
Though looking more into `StatePreparation`, they specify `num_qubits` as optional.
https://github.com/Qiskit/qiskit/blob/a7cc16624a719332b8ad5ab764425dca33f3952c/qiskit/circuit/library/data_preparation/state_preparation.py#L45
Then they just get `num_qubits` through the ` _get_num_qubits method`.
https://github.com/Qiskit/qiskit/blob/a7cc16624a719332b8ad5ab764425dca33f3952c/qiskit/circuit/library/data_preparation/state_preparation.py#L111

Hirmay committed:
Update test_gate_definitions.py

woodsp-ibm commented:
If StatePreparation is excluded from that test it made sense to me that your new class probably should be too.

Yes StatePreparation has num_qubits as optional and does that computation you refer to. It seemed like this could be done the same if that seems reasonable - it would perhaps make them more consistent in that regard.

Hirmay commented:
Okay, that makes sense. I will try to do that as well. Thank you so much, Steve, for all the help and support!

woodsp-ibm commented on qiskit/circuit/library/data_preparation/state_preparation.py:
```diff
@@ -414,3 +439,131 @@ def _multiplex(self, target_gate, list_of_angles, last_cnot=True):
+            A positive integer representing the number of qubits used.
+
+    Returns:
+        cirq.Circuit: A quantum circuit that creates the uniform superposition state: $\frac{1}{\sqrt{M}} \sum_{j=0}^{M-1}  \ket{j} $. 
```
I think its more normal to have the Args and descriptions thereof just in the init doc string, like StatePreparation does. That will hopefully also take care of the return type here!!!

Hirmay replied:
Thanks for the suggestion, Steve. I just pushed your recommendation.

woodsp-ibm replied:
I should have looked more carefully - you have `M` as a public instance variable which is maybe why you have it in Args there. I am more used to seeing the public instance vars done under Attributes there. I guess you want that as a public var? I am noting that num_qubits is a private var though. StatePreparation they are all marked private i.e. a leading underscore in the name.

Hirmay replied:
You're right, there is not a strong argument to have `M` as a public instance variable. I have made it as a private var.

Hirmay committed:
made the format consistent with StatePreparation class

woodsp-ibm commented on qiskit/circuit/library/data_preparation/state_preparation.py:
```diff
@@ -414,3 +439,131 @@ def _multiplex(self, target_gate, list_of_angles, last_cnot=True):
+            num_qubits (int): The number of qubits used.
+        """
+        super().__init__("USup", num_qubits, [])
+        assert (
```
If you look at StatePreparation this type of check raises an error rather than doing and assert. I think assert if not used in Qiskit rather it does some checks and raises and error in generaI. Here I would check and raise a ValueError rather than being an assert - my take is that would be more standard than the QiskitError that StatePreparation uses.

Your unit tests should check these error paths and make sure an error is raised when expected.

Hirmay replied:
You're right, I have replaced `assert` with `raise ValueError`. I will look into unit tests too. Thanks, again!

Hirmay replied:
I have added two test cases, hopefully everything is all well now. Thank you, Steve, for everything!

Hirmay committed:
Put description and arguments in init

Hirmay committed:
replaced assert with raise ValueError

Hirmay committed:
small mistake in Returns

Hirmay committed:
added test cases for ValueError cases

Hirmay commented:
@woodsp-ibm I think the reason why it is still giving the same error in Coverage report test is it might be testing it without the updated `TestGateEquivalenceEqual` which excludes this class, whereas the old one doesn't. One solution is to just submit a PR with the updated  `TestGateEquivalenceEqual` and once it is merged try this, but that may not be preferable with the reviewers.

woodsp-ibm commented:
Its not failing the same way though
```
test.python.circuit.library.test_generalized_uniform_superposition_gate.TestGeneralizedUniformSuperposition.test_generalized_uniform_superposition_gate
-------------------------------------------------------------------------------------------------------------------------------------------------------

Captured traceback:
~~~~~~~~~~~~~~~~~~~
    Traceback (most recent call last):

      File "/home/runner/work/qiskit/qiskit/test/python/circuit/library/test_generalized_uniform_superposition_gate.py", line 38, in test_generalized_uniform_superposition_gate
    gate = Generalized_Uniform_Superposition_Gate(M, n)

      File "/home/runner/work/qiskit/qiskit/qiskit/circuit/library/data_preparation/state_preparation.py", line 484, in __init__
    super().__init__("USup", num_qubits, [])

      File "/home/runner/work/qiskit/qiskit/qiskit/circuit/gate.py", line 46, in __init__
    super().__init__(name, num_qubits, 0, params, label=label, duration=duration, unit=unit)

      File "/home/runner/work/qiskit/qiskit/qiskit/circuit/instruction.py", line 80, in __init__
    raise CircuitError("num_qubits and num_clbits must be integer.")

    qiskit.circuit.exceptions.CircuitError: 'num_qubits and num_clbits must be integer.'


test.python.circuit.library.test_generalized_uniform_superposition_gate.TestGeneralizedUniformSuperposition.test_incompatible_M
```
this is from your code. In StatePreparation the super.init call is made later once the number of qubits is computed. Maybe compare to that as I think its failing due the super init call now. Most likely you end up passing None now because its optional so it needs to be done once you compute the value like StatePreparation does

Hirmay committed:
Update state_preparation.py

Hirmay commented:
Yes, Steve, the error is different, but only slightly. The fundamental error is the same. I think the error coming from `ddt` is not there because of the decorator I added later (which didn't go under checks, since I was constantly updating the files :sweat_smile: ). I believe if the same is checked with the class being excluded, this error won't be produced, and this is the only primary error currently.

woodsp-ibm commented on qiskit/circuit/library/data_preparation/state_preparation.py:
```diff
@@ -414,3 +439,137 @@ def _multiplex(self, target_gate, list_of_angles, last_cnot=True):
+        self,
+        params: Union[str, list, int, Statevector],
+        M : int = 2,
+        num_qubits: Optional[int] = None,
```
I am not understanding why params is here. I know they are passed to the parent but this is not something that is used here right - your only params are M. In your test, where you are doing `Generalized_Uniform_Superposition_Gate(M, n)` the M there goes to params (the first parameter in the list here), n ends up in M (the 2nd position) and num_qubits is left as None.

Hence this call fails in the parent `super().__init__("USup", num_qubits, [])` as it passes None (as far as I can see).

My take would be to remove params from here, have M first here, not have a default value so you have to provide a value, and then num_qubits as you have it.

Next move this call `super().__init__("USup", num_qubits, [])` to the end of the method and change it to `super().__init__("USup", self._num_qubits, [M])` This way you always pass a valid value of num qubits to the parent since you compute it if its None based on M. Then M is made  single value in a list passed as parameters to the parent - this is kinda the equivalent of what params is to the StatePreparation. You can see how its handled there.

As to ddt you are not using it here. You can parameterize the test and even create the data so you do not need a for loop rather ddt will create a single test for each data entry there. That is better as all the tests will always run - in the for loop, as you have it, it will stop at the first failure. Usually we like to run all tests and if a test had independent parts like this, but is in one test, it would be done with the self.subTest. This is something that can be sorted once the code is running properly - I think what I am suggesting above will fix things.

Hirmay committed:
incorporate Steve's helpful suggestions

Hirmay committed:
small bug

Hirmay committed:
test_case fix

Hirmay reviewed (commented):
Yes, you're correct. I have incorporated the changes you suggested. Hopefully it passes the checks. I'll later look into ddt, once the code is running properly. Thank you so much, Steve!

Hirmay committed:
function for returning M

Hirmay committed:
oops..forgot "" marks

Hirmay committed:
Update state_preparation.py

woodsp-ibm commented on qiskit/circuit/library/data_preparation/state_preparation.py:
```diff
@@ -414,3 +439,143 @@ def _multiplex(self, target_gate, list_of_angles, last_cnot=True):
+    
+    def _get_num_qubits(self):
+        """Returns the number of qubits used by the gate."""
+        return self._num_qubits
```
I don't think either of these private methods are used/needed right. As they simply just return an instance var value using that directly internally, as is done, is fine (and it saves the overhead of a method call)

Hirmay replied:
Yes, Steve, I'll remove them.

woodsp-ibm commented on qiskit/circuit/library/data_preparation/state_preparation.py:
```diff
@@ -414,3 +439,143 @@ def _multiplex(self, target_gate, list_of_angles, last_cnot=True):
+                "M must be a positive integer greater than 1."
+            )
+        if num_qubits is None:
+            num_qubits = np.ceil(np.log2(M))
```
Make sure you add a unit test that has the num_qubits as None so it tests this. I am not sure whether this returns an int type exactly - I know some instance checks we had for Python int etc would fail - in places I have seen the return from ceil cast to int so I am not sure. It may fail in the super class when this is passed to it in the super init call.

Hirmay replied:
Oh wow, there is a case when `M` is a power of 2 and that would not be handled well. Thanks for pointing it out, I'll include the test case when `num_qubits = None` .

Hirmay committed:
removed get methods for num_qubit and M

Hirmay committed:
added power of 2 condition

Hirmay committed:
included test function when num_qubits is None

woodsp-ibm commented:
Looks like things are passing - except you need to run `black` on the code to make sure its formatted with that as presently the black check is failing saying 3 files need formatting.

Hirmay committed:
blacked init

Hirmay committed:
blacked state_prep

Hirmay committed:
blacked test_generalized_uniform_superposition_gate.py

Hirmay commented:
Yes, Steve, I have just done that. Hopefully, now all tests should pass. Thank you so you much for painstakingly helping me to debug this code, I couldn't have done it without your help and support.

Hirmay commented:
I don't why these two files are failing black's check, when locally they're passing. I have also manually reviewed changes and there doesn't seem to be any.
![Screenshot from 2024-04-10 21-17-42](https://github.com/Qiskit/qiskit/assets/56473003/6c4aab92-32f7-435b-b0f0-96c01d10fc30)

woodsp-ibm commented:
Could you be using an older version of black? Maybe try updating it - CI will be using the latest version as I do not see any restriction in requirements_dev.txt other than avoiding one with a known problem

Hirmay committed:
reblacked state_prep

Hirmay committed:
reblacked test_generalized_uniform_superposition_gate.py

Hirmay commented:
I changed to an environment with a newer version of python (since in that environment black was updated to the 'latest' version) and that gave some minor changes to these two files. Hopefully it passes this check now.

woodsp-ibm commented:
Its just the one file now. You have the pyproject.toml file there from the qiskit root right - that limits line length to 100. I see lines in the docstring that seem longer than that. Perhaps thats whys its failing?

Hirmay commented:
Could be, since I'm on Python version 3.12.2 (older environment was on Python 3.9.12) . Now it could be that the checker might be running on an older version (between 3.9 and 3.12) which may cause this discrepancy.

Hirmay committed:
shorterned max line length

Hirmay committed:
reblacked state_prep

Hirmay commented:
Hi Steve, I downgraded to Python 3.8.18 (one which the checker uses), ran black, and then replace state_preparation.py file. This somehow still gives error. One possible reason would be that the black used by the checker is not 24.3.0, but rather some old version. Could you please check which version it could be, then I just have to install it and it would possibly pass the test?

woodsp-ibm commented:
I downloaded your state_preparation and put in in my qiskit project (I just put it in `qiskit` folder to do black so it was easy to remove without affecting the orginal one). Then from the root of that project I.e the level above qiskit I ran `black --check qiskit test tools examples setup.py` and it said that file would be changed. I removed --check and sure enough it changed it as I did a diff with the same file still in the Downloads folder I had made the copy into qiskit. My local Python is 3.9.7 and I updated black to the latest version. Is this how you are doing things - ie run the black from the project root? I am not sure why you are seeing a different outcome locally.

Looking at the above where you show me its says the files would be unchanged I just see 3. So maybe you are putting them elsewhere. Leave them in the project in the place where there are in the clone of the repo and run the above. I see something like over 1200 files are unchanged if I do the check. If you run them and it changes one you can just commit the new changed files(s) from right there.

Hirmay commented:
Hi Steve, I am just checking that particular file, since I thought it won't make a difference. If you don't mind, can you please replace the older file with your updated file? That would resolve this matter.

Hirmay committed:
reblacked state_prep

woodsp-ibm commented:
I see you managed to get black to alter the file so its ok in CI here. The changes look like what I saw locally. I see now its passing black that its getting further but now failing with the lint.

Hirmay commented:
Yes, Steve, two steps forward, one step backward :sweat_smile:. I'll hopefully fix the code according to lint's remarks and then maybe all checks would be successful.

Hirmay committed:
for pyline

Hirmay committed:
pylinted state_preparation.py

Hirmay committed:
pylinted test_generalized_uniform_superposition_gate.py

Hirmay committed:
pylinted test_gate_definitions.py

Hirmay committed:
pylinted test_generalized_uniform_superposition_gate.py

Hirmay commented:
Hi Steve, finally all the checks have been successful :smile:. Thank you for all the help and support. Let me know what are the next future steps? Should I now select Ready for review?

woodsp-ibm commented:
> Hi Steve, finally all the checks have been successful 😄. Thank you for all the help and support. Let me know what are the next future steps? Should I now select Ready for review?

You are most welcome. Yes I would take it out of Draft mode - it will eventually need a reno release note to be eligible to be merged (see https://github.com/Qiskit/qiskit/blob/main/CONTRIBUTING.md#adding-a-new-release-note ) - this documents  user facing changes, like this new feature. You can add it now if you like - there are examples in the releasenotes folder - or wait until someone from the core qiskit team has had a chance to go over this - up to you.

Hirmay commented:
Okay, I'm taking it out of Draft mode. Yes, I'll later write a reno release note. Thanks, again, Steve!

Hirmay committed:
Added GeneralizedUniformSuperposition gate

Added GeneralizedUniformSuperposition gate class to the StatePreparation file in Circuit library.

Hirmay commented:
I have added the reno release note as well. Hopefully this pull request gets approved.

Hirmay committed:
modified:   releasenotes/notes/generalized-uniform-superposition-gate-3bd95ffdf05ef18c.yaml

woodsp-ibm commented on releasenotes/notes/generalized-uniform-superposition-gate-3bd95ffdf05ef18c.yaml:
```diff
@@ -15,20 +15,8 @@ features:
-fixes:- |
-     Refer to
-    `#11725 <https://github.com/Qiskit/qiskit/issues/11725>` for more
+fixes:
```
I would remove the fixes section - that is more for bug fixes. You added a new feature so it just needs the features section which you have. You can delete the prelude section too so that all this contains is one section. As to the [SV24] in the features maybe you can state the reference there more directly inline in the text.

Hirmay replied:
Okay, Steve, I'll do that.

woodsp-ibm commented:
A more minor thing - maybe you'd like to edit the title of the PR to something like `Add Generalized Uniform State Preparation` or something that is more descriptive of what this adds as `Update state_preparation.py` is more about how you did it and really says nothing about what the update is, plus people can easily see that your new feature was added there from the file changes.

Hirmay committed:
Updated release notes based on Steve's suggestions

Hirmay commented:
Yes, that makes a lot of sense, Steve. I have changed the title. I hope all things have been checked out now and only have to wait for the reviewers to see this. Once again, thank you so much for all the help and support. This felt like a mini QAMP session :smile:.

woodsp-ibm commented:
@jakelishman We had a discussion in the original issue #11735 that was done for this feature. I have tried to help out here. Its in a state now I think where its ready for someone on the qiskit team to review

Hirmay committed:
Merge branch 'main' into patch-2

Hirmay committed:
Update release note

Hirmay commented:
Hi Steve, do you know on average how long does it take for the pull-request to be reviewed?

jakelishman commented on qiskit/circuit/library/data_preparation/state_preparation.py:
```diff
@@ -414,3 +419,139 @@ def _multiplex(self, target_gate, list_of_angles, last_cnot=True):
+    """
+    Class that implements a generalized uniform superposition state, using n qubits,
+    following the Shukla-Vedula algorithm [SV24].
+    """
```
I'm not sure that the word "generalised" is improving the description here: is there something that's more general than a standard "uniform superposition"

$$
\frac{1}{\sqrt M}\sum_{m=0}^{M-1} \lvert m\rangle_n?
$$

Perhaps we could call this `UniformSuperpositionGate` which is a bit less typing for people?

jakelishman commented on qiskit/circuit/library/data_preparation/state_preparation.py:
```diff
@@ -414,3 +419,139 @@ def _multiplex(self, target_gate, list_of_angles, last_cnot=True):
+                states have zero amplitudes. Here M need not be an integer power of 2.
+
+            num_qubits (int):
+                A positive integer representing the number of qubits used.
```
The idea of the no-single-letters lint is to use descriptive names for variables.  I think `m_value` might be more descriptive as `num_elements`?

jakelishman commented on qiskit/circuit/library/data_preparation/state_preparation.py:
```diff
@@ -414,3 +419,139 @@ def _multiplex(self, target_gate, list_of_angles, last_cnot=True):
+
+        Returns:
+            A quantum circuit that creates the uniform superposition state:
+            $\frac{1}{\sqrt{M}} \sum_{j=0}^{M-1}  \ket{j} $.
```
An `__init__` method doesn't actually return anything.  This little bit of description might be better up in the class docstring.

It would also be good to comment on the assumptions of the _input_ state as well; the gate doesn't create the uniform superposition state from any input state (that would be a violation of unitarity), it only creates the state from an input of $\lvert 0\cdots0\rangle$.  That's totally fine, but we should comment that the effect on other states is deterministic but undefined.

jakelishman commented on qiskit/circuit/library/data_preparation/state_preparation.py:
```diff
@@ -414,3 +419,139 @@ def _multiplex(self, target_gate, list_of_angles, last_cnot=True):
+            23(38): pp. 1-32 (2024).
+
+        Raises:
+            QiskitError: ``num_qubits`` parameter used when ``params`` is not an integer
```
I'm not sure what `params` is here?  I think this might be a little out-of-date, because nothing appears to raise `QiskitError` either.

jakelishman commented on qiskit/circuit/library/data_preparation/state_preparation.py:
```diff
@@ -414,3 +419,139 @@ def _multiplex(self, target_gate, list_of_angles, last_cnot=True):
+            $\frac{1}{\sqrt{M}} \sum_{j=0}^{M-1}  \ket{j} $. It needs only
+            $O(\log_2 (M))$ qubits and $O(\log_2 (M))$ gates, hence providing an
+            exponential improvement, in terms of reduced resources and complexity,
+            compared to alternative methods in existing literature.
```
The best way of making a "note" in rST syntax is to do
```rst
.. note::

    The Shukla-Vedula algorithm ...
```

This would be a little better in the main class docstring (since it's not related to the _initialisation_ of the object).  We shouldn't compare it to "existing literature", because this paper _is_ "existing literature" (and we don't really need to advertise the quality of research papers) - it's good / useful to mention that it's a more efficient creation than using the more general `StatePreparation` though.

jakelishman commented on qiskit/circuit/library/data_preparation/state_preparation.py:
```diff
@@ -414,3 +419,139 @@ def _multiplex(self, target_gate, list_of_angles, last_cnot=True):
+            if (m_value & (m_value - 1)) == 0:  # If m_value is an integer power of 2
+                num_qubits = int(np.log2(m_value))
+            else:  # If m_value is not an integer power of 2
+                num_qubits = int(np.ceil(np.log2(m_value)))
```
It's better to use `log2` (and all other maths functions) from the Python module `math` than `numpy` when we're only acting on a single scalar, because there's a lot less overhead.

jakelishman commented on qiskit/circuit/library/data_preparation/state_preparation.py:
```diff
@@ -414,3 +419,139 @@ def _multiplex(self, target_gate, list_of_angles, last_cnot=True):
+                )
+        self._num_qubits = num_qubits
+        self._m_value = m_value
+        super().__init__("USup", self._num_qubits, [m_value])
```
There's no need to store `_num_qubits` manually; it's part of the regular gate structure and accessible as `self.num_qubits` after the `super().__init__` call.  We also don't need to store `_m_value` both manually _and_ in `self.params` - it's best just to leave it in `self.params`, most likely.

jakelishman commented on qiskit/circuit/library/data_preparation/state_preparation.py:
```diff
@@ -414,3 +419,139 @@ def _multiplex(self, target_gate, list_of_angles, last_cnot=True):
+
+        Returns:
+            QuantumCircuit: The quantum circuit implementing the gate.
+        """
```
`_define` methods shouldn't return anything, they should set `self._definition` to the quantum circuit only.  Please could you have a look at how some other gates are implemented and make sure they're similar?

Cryoris replied:
@Hirmay I think your function is already correctly implemented in that regard, just the docstring says a circuit is returned, even though it doesn't. Maybe you could just delete the entire docstring here.

jakelishman commented on qiskit/circuit/library/data_preparation/state_preparation.py:
```diff
@@ -414,3 +419,139 @@ def _multiplex(self, target_gate, list_of_angles, last_cnot=True):
+    def to_instruction(self):
+        """Returns the gate instruction."""
+        qc = self._define()
+        return qc.to_instruction()
```
If your gate is set up correctly, all of these methods will be automatically inherited from `Gate`.  The reason you're finding that you need to define some of them is because your `_define` isn't quite doing the right thing.

We need to fix `_define` to set `self._definition` and then remove these.

jakelishman commented on qiskit/circuit/library/data_preparation/state_preparation.py:
```diff
@@ -414,3 +419,139 @@ def _multiplex(self, target_gate, list_of_angles, last_cnot=True):
+        instruction = self.to_instruction()
+        qc = QuantumCircuit(self._num_qubits)
+        qc.append(instruction, list(range(self._num_qubits)))
+        return np.array(Operator(qc).data)
```
We can remove this method - `Gate` already defines a `to_matrix` method, which is the standard, and that uses `__array__`.  If you want to provide a matrix method, we should define `__array__` (again, please see other implementations for examples), and remove this method.

jakelishman commented on test/python/circuit/library/test_generalized_uniform_superposition_gate.py:
```diff
@@ -0,0 +1,82 @@
+            if (m_value & (m_value - 1)) == 0:  # If m_value is an integer power of 2
+                n = int(np.log2(m_value))
+            else:  # If m_value is not an integer power of 2
+                n = int(np.ceil(np.log2(m_value)))
```
The `if/else` is not really necessary here - the ceiling of a value that's already integral is just itself.

jakelishman commented on test/python/circuit/library/test_generalized_uniform_superposition_gate.py:
```diff
@@ -0,0 +1,82 @@
+            gate = GeneralizedUniformSuperpositionGate(m_value, n)
+            unitary_matrix = np.real(gate.to_unitary())
+            actual_sv = unitary_matrix[:, 0]
+            self.assertTrue(np.allclose(desired_sv, actual_sv))
```
Numpy provides a method called `np.testing.assert_allclose`, which is a bit better to use in tests than `assertTrue(np.allclose(x, y))` because we get better error messages when it fails.

jakelishman commented on releasenotes/notes/generalized-uniform-superposition-gate-3bd95ffdf05ef18c.yaml:
```diff
@@ -0,0 +1,24 @@
+    usp_gate = GeneralizedUniformSuperpositionGate(M, n)
+    qc = QuantumCircuit(n)
+    qc.append(usp_gate, list(range(n)))
+    qc.draw('mpl',style="clifford")
```
Code blocks need to be indented, but also, since this one is trying to draw something with Matplotlib, it ought to be a `.. plot::` rather than a `.. code-block::`.

jakelishman commented on test/python/circuit/library/test_generalized_uniform_superposition_gate.py:
Also in this test file, please can we check a couple of other standard behaviours of gates:

- it can be appended to a circuit, and then we can call `Operator(qc)` on the circuit and get the correct output
- it can be appended to a circuit, then the result can be transpiled to some simple basis set (you can use the basis gates you use in the definition, that's fine), and the output is valid
- we can call `.inverse` on the gate and it works.  We might also want to test that `qc.append(gate, [...]); qc.append(gate.inverse(), [...])` produces the identity matrix under `Operator`.

We don't need to check 80 different numbers of elements for each of these - testing just a couple, such as a one-element, four-element and (say) 13-element superpositions would be fine.

It would be a bit better to use `ddt` test parametrisation (see other tests in the same folders for example) than using `for` loops inside the tests, so that on failures, it's clearer _which_ tests are failing, and all the tests can run regardless of an earlier part of the loop fails.

jakelishman reviewed (commented):
Thank you for sticking with this.  Please can you also try building the documentation and making sure that it all looks like it's supposed to?  It looks to me that there might be a few places where it's not working the way we expect.

It might be good to have another look at how things like `StatePreparation` implement all the public methods that they implement, and make sure that the methods we've implemented here work in the same way - there's a few in here that are not quite doing what they're meant to do.

jakelishman commented:
Also, Steve - thanks very very much for doing all the first parts of the review.

Hirmay committed:
Merge branch 'main' into patch-2

Hirmay committed:
implemented the changes

Hirmay committed:
fixed error

Hirmay committed:
Update test_uniform_superposition_gate.py

Hirmay committed:
Update uniform-superposition-gate-3bd95ffdf05ef18c.yaml

Hirmay committed:
Update uniform-superposition-gate-3bd95ffdf05ef18c.yaml

Hirmay commented:
Hi @jakelishman, thank you so much for the valuable feedback. I have incorporated mostly all the changes you asked for. I changed it to `.. plot::` but since that was giving some error (unable to tell because `-q` flag, I reverted back to `.. code-block::`. And I changed the last line to just `qc.draw()`, so hopefully now that is also fine. 

These are the following changes done:
1. Change name from "GenralizedUniformSuperpositionGate" to "UniformSuperpositionGate" as suggested.
2. Replace the following files to remove reference to "generalized" (as suggested): \_\_init__.py, state_preparation.py, test_gate_definitions.py. 
3. Replace "test_generalized_uniform_superposition_gate.py" with "test_uniform_superposition_gate.py". 
4. Replace "generalized-uniform-superposition-gate-3bd95ffdf05ef18c.yaml" with "uniform-superposition-gate-3bd95ffdf05ef18c.yaml"
5. Note: If/else condition has been changed (as suggested).
6. Included math.ceil and math.log2 instead of np.ceil and np.log2.
7. m_value replaced by num_superpos_states
8. Changed assertTrue(np.allclose(x, y)) to np.testing.assert_allclose (as suggested).
9. Changed QiskitError to ValueError in the docstring of state_preparation.py file.
10. Deleted to_unitary() function in state_preparation.py and moved it to test_uniform_superposition_gate.py
11. Removed redundant definitions.
12. Addressed the following comment: "There's no need to store \_num_qubits manually; it's part of the regular gate structure and accessible as self.num_qubits after the super().\_\_init__ call. We also don't need to store _m_value both manually and in self.params - it's best just to leave it in self.params, most likely."    

Please let me know if there are any more changes required. Thanks in advance!

Hirmay commented:
Hi @jakelishman, just a friendly reminder regarding this.

Hirmay commented:
Hi @jakelishman, can you please take a look at the changes? If you're busy with other things, please let me know.

Hirmay commented:
Hi @jakelishman or @woodsp-ibm, can you please review the recommended changes?

eliarbel commented:
> Hi @jakelishman or @woodsp-ibm, can you please review the recommended changes?

woodsp-ibm commented:
@Hirmay In the end here you need a maintainer of Qiskit to approve this so it can be merged. As you had initially brought up the idea on algorithms repo, where I am a maintainer, and I had suggested doing it here as it seemed more general, I had provided some initial help/guidance to get things in a reasonable shape as the person who we thought might have helped you out was on extended leave, so I figured I could at least help get things going. But for final "polishing" and acceptance the approval needs to come from a Qsikit maintainer who better know internals/details of how things ought to be so my help could only go so far.

Hirmay commented:
Hi Steve, (as stated in qiskit's slack) I didn't know that, thanks for informing me about that. I again want to state that I'm quite grateful to you for going above and beyond in the help and guidance you provided throughout this ardous polishing process.

Cryoris commented on qiskit/circuit/library/data_preparation/state_preparation.py:
```diff
@@ -414,3 +418,107 @@ def _multiplex(self, target_gate, list_of_angles, last_cnot=True):
+    A. Shukla and P. Vedula, ``An efficient quantum algorithm for preparation
+    of uniform quantum superposition states,`` Quantum Information Processing,
+    23(38): pp. 1-32 (2024).
+    """
```
Here's some suggestions to make the docstring more consistent with the rest of Qiskit and fix the math display
```suggestion
    r"""Implements a uniform superposition state.

    This gate is used to create the uniform superposition state
    :math:`\frac{1}{\sqrt{M}} \sum_{j=0}^{M-1}  |j\rangle` when it acts on an input
    state :math:`|0...0\rangle`. Note, that `M` is not required to be 
    a power of 2, in which case the uniform superposition could be 
    prepared by a single layer of Hadamard gates.

    .. note::

        This class uses the Shukla-Vedula algorithm [1], which only needs
        :math:`O(\log_2 (M))` qubits and :math:`O(\log_2 (M))` gates, 
        to prepare the superposition.

    **References:**
    [1]: A. Shukla and P. Vedula (2024), An efficient quantum algorithm for preparation
    of uniform quantum superposition states, `Quantum Inf Process 23, 38
    <https://link.springer.com/article/10.1007/s11128-024-04258-4>`_.
    """
```

Cryoris commented on qiskit/circuit/library/data_preparation/state_preparation.py:
```diff
@@ -414,3 +418,107 @@ def _multiplex(self, target_gate, list_of_angles, last_cnot=True):
+    ):
+        r"""
+        Initializes UniformSuperpositionGate.
+
```
This first line is not really needed if it doesn't mention any special behavior 
```suggestion
        r"""
```

1ucian0 commented:
> I have incorporated mostly all the changes you asked for.

For next time, it is easier to track changes if you reply in the conversation for each code comment.

1ucian0 commented on releasenotes/notes/uniform-superposition-gate-3bd95ffdf05ef18c.yaml:
```diff
@@ -0,0 +1,23 @@
+---
+features:
+  - |
+    Implemented UniformSuperpositionGate class, which allows 
```
```suggestion
    Implemented :class:`.UniformSuperpositionGate` class, which allows 

```

Hirmay replied:
Okay, will do that! Thanks!

1ucian0 commented on releasenotes/notes/uniform-superposition-gate-3bd95ffdf05ef18c.yaml:
```diff
@@ -0,0 +1,23 @@
+       n = 3
+       usp_gate = UniformSuperpositionGate(M, n)
+       qc = QuantumCircuit(n)
+       qc.append(usp_gate, list(range(n)))
```
`M` is explained in the text, so it is good. However, many times it is better to be more explicit with name variables:
```suggestion
       num_qubits = 3
       usp_gate = UniformSuperpositionGate(M, num_qubits)
       qc = QuantumCircuit(num_qubits)
       qc.append(usp_gate, list(range(num_qubits)))

```

Cryoris commented on qiskit/circuit/library/data_preparation/state_preparation.py:
```diff
@@ -414,3 +418,107 @@ def _multiplex(self, target_gate, list_of_angles, last_cnot=True):
+    def __init__(
+        self,
+        num_superpos_states: int = 2,
+        num_qubits: Optional[int] = None,
```
With only Python 3.7+ support now we can use the `|` operand for typehints 🙂 
```suggestion
        num_qubits: int | None = None,
```

Cryoris commented on qiskit/circuit/library/data_preparation/state_preparation.py:
```diff
@@ -414,3 +418,107 @@ def _multiplex(self, target_gate, list_of_angles, last_cnot=True):
+                basis states with an amplitude of 1/sqrt(M) in the uniform superposition
+                state ($\frac{1}{\sqrt{M}} \sum_{j=0}^{M-1}  \ket{j} $, where
+                $1< M <= 2^n$). Note that the remaining (2^n - M) computational basis
+                states have zero amplitudes. Here M need not be an integer power of 2.
```
Could you update the math using
```
:math:`<math goes here>`
```
as I proposed in the comment above? 🙂

Hirmay replied:
Yes, it is interesting to see a new way of writing math equations, thanks!

Cryoris commented on releasenotes/notes/uniform-superposition-gate-3bd95ffdf05ef18c.yaml:
```diff
@@ -0,0 +1,23 @@
+---
+features:
+  - |
+    Implemented UniformSuperpositionGate class, which allows 
```
This allows Sphinx to link the text to the class docs
```suggestion
    Implemented :class:`.UniformSuperpositionGate` class, which allows 

```

Cryoris commented on releasenotes/notes/uniform-superposition-gate-3bd95ffdf05ef18c.yaml:
```diff
@@ -0,0 +1,23 @@
+    the creation of a uniform superposition state using 
+    the Shukla-Vedula algorithm [SV24]. This feature facilitates the 
+    creation of quantum circuits that produce a uniform superposition 
+    state $\frac{1}{\sqrt{M}} \sum_{j=0}^{M-1}  \ket{j}$, where $M$ 
```
See the comments about math font above 🙂

Hirmay replied:
Yes, will replace with that notation. Thanks!

Cryoris commented on test/python/circuit/test_uniform_superposition_gate.py:
```diff
@@ -0,0 +1,139 @@
+class TestUniformSuperposition(QiskitTestCase):
+    """Test initialization with UniformSuperpositionGate class"""
+
+    @data(2, 3, 5, 13)
```
I think we can remove the 13 qubit case, it seems computationally heavy for a test, without testing behavior that is different from the smaller qubit cases. This way we can keep the tests fast.

Hirmay replied:
That makes sense, wasn't thinking that these tests might be ran again and again, so every second matters.

1ucian0 commented on releasenotes/notes/uniform-superposition-gate-3bd95ffdf05ef18c.yaml:
```diff
@@ -0,0 +1,23 @@
+  - |
+    Implemented UniformSuperpositionGate class, which allows 
+    the creation of a uniform superposition state using 
+    the Shukla-Vedula algorithm [SV24]. This feature facilitates the 
```
Does references worked from a different doc?

Hirmay replied:
Can you please elaborate on this?

1ucian0 replied:
`SV24` is a reference defined in the docstring. I am not sure if it can be resolved (maybe it can)

Hirmay replied:
Hi @1ucian0, I have removed `[SV24]` since it does not add anything significant to the note and seems to be the most efficient way to resolve the issue. Please take a look at all the changes and let me know if you need anything else from my side.

Hirmay committed:
Update qiskit/circuit/library/data_preparation/state_preparation.py

Sounds good!

Co-authored-by: Julien Gacon <gaconju@gmail.com>

Hirmay committed:
Update qiskit/circuit/library/data_preparation/state_preparation.py

Okay!

Co-authored-by: Julien Gacon <gaconju@gmail.com>

Hirmay committed:
Update qiskit/circuit/library/data_preparation/state_preparation.py

oh that's interesting...I didn't know that. Thanks for the info!

Co-authored-by: Julien Gacon <gaconju@gmail.com>

Hirmay committed:
Update releasenotes/notes/uniform-superposition-gate-3bd95ffdf05ef18c.yaml

Ahhh...that's nice!

Co-authored-by: Julien Gacon <gaconju@gmail.com>

Hirmay committed:
Update releasenotes/notes/uniform-superposition-gate-3bd95ffdf05ef18c.yaml

You're quite right, will help for others who might look at the code in future.

Co-authored-by: Luciano Bello <bel@zurich.ibm.com>

Hirmay committed:
Update state_preparation.py

incorporated Julien's changes!

Hirmay committed:
Update test_uniform_superposition_gate.py

Incorporated Julien's optimization suggestion.

Hirmay committed:
Update uniform-superposition-gate-3bd95ffdf05ef18c.yaml

implemented both reviewers' suggestions.

Hirmay commented:
I have added almost every suggestion (handled even @1ucian0 's last suggestion), if you get time, please take a look at this. Thanks in advance XD!

Hirmay committed:
Update uniform-superposition-gate-3bd95ffdf05ef18c.yaml

removed SV24

1ucian0 committed:
Merge branch 'main' into patch-2

Hirmay committed:
Update state_preparation.py

Cryoris commented on qiskit/circuit/library/data_preparation/state_preparation.py:
```diff
@@ -240,3 +242,98 @@ def validate_parameter(self, parameter):
+            num_superpos_states (int):
+                A positive integer M = num_superpos_states (> 1) representing the number of computational
+                basis states with an amplitude of 1/sqrt(M) in the uniform superposition
+                state (:math:`\frac{1}{\sqrt{M}} \sum_{j=0}^{M-1}  \ket{j}`, where
```
`\ket` is not supported in the docs build, you can use `|j\rangle` instead, like you did above. There's also one more occurrence in the releasenotes 🙂

Hirmay replied:
Okay, that makes sense, I have done that.

Cryoris commented on qiskit/circuit/library/data_preparation/state_preparation.py:
```diff
@@ -240,3 +242,98 @@ def validate_parameter(self, parameter):
+                A positive integer M = num_superpos_states (> 1) representing the number of computational
+                basis states with an amplitude of 1/sqrt(M) in the uniform superposition
+                state (:math:`\frac{1}{\sqrt{M}} \sum_{j=0}^{M-1}  \ket{j}`, where
+                :math:`1< M <= 2^n`). Note that the remaining (2^n - M) computational basis
```
Did you want to put these in math mode as well?
```suggestion
                :math:`1< M <= 2^n`). Note that the remaining (:math:`2^n - M`) computational basis
```

Hirmay replied:
Okay, done.

Cryoris commented on qiskit/circuit/library/data_preparation/state_preparation.py:
```diff
@@ -240,3 +242,98 @@ def validate_parameter(self, parameter):
+
+        """
+        if not (isinstance(num_superpos_states, int) and (num_superpos_states > 1)):
+            raise ValueError("num_superpos_states must be a positive integer greater than 1.")
```
We assume users obey the type hints, so you can remove that initial check 🙂 
```suggestion
        if not num_superpos_states > 1:
            raise ValueError("num_superpos_states must be a positive integer greater than 1.")
```

Hirmay replied:
Okay, no that seems fair. I have done this change.

Cryoris commented on qiskit/circuit/library/data_preparation/state_preparation.py:
```diff
@@ -240,3 +242,98 @@ def validate_parameter(self, parameter):
+                states have zero amplitudes. Here M need not be an integer power of 2.
+
+            num_qubits (int):
+                A positive integer representing the number of qubits used.
```
This should explain what happens if the value is `None`

Hirmay replied:
Hmmm...I have added this!

Cryoris commented on qiskit/circuit/library/data_preparation/state_preparation.py:
```diff
@@ -240,3 +242,98 @@ def validate_parameter(self, parameter):
+        """
+        Defines the gate operation.
+
+        """
```
This is clear from the method name, in this case we typically omit the docstring
```suggestion
```

Hirmay replied:
Okay, got it.

Cryoris commented on qiskit/circuit/library/data_preparation/state_preparation.py:
```diff
@@ -240,3 +242,98 @@ def validate_parameter(self, parameter):
+        Defines the gate operation.
+
+        """
+        qreg = QuantumRegister(self._num_qubits, "q")
```
Could you remove the usage of `qreg` (in particular since it is not tied to the circuit)? You can just use integer indices and ranges instead. For example
* `qc.x(idx)` instead of `qc.x(qreg[idx])`
* `qc.h(range(m))` instead of `qc.h(qreg[0:m])`

Hirmay replied:
Okay, julien, I have updated it based on the example provided. Thanks!

Cryoris commented on qiskit/circuit/library/data_preparation/state_preparation.py:
```diff
@@ -240,3 +242,98 @@ def validate_parameter(self, parameter):
+            self.definition = qc
+            return
+
+        n_value = [int(x) for x in list(np.binary_repr(num_superpos_states))][::-1]
```
How about something like
```suggestion
        n_value = [int(x) for x in reversed(np.binary_repr(num_superpos_states))]
```
This should allow to directly iterate in reverse instead of creating the list twice.

Hirmay replied:
I have done this, thanks!

Cryoris commented on qiskit/circuit/library/data_preparation/state_preparation.py:
```diff
@@ -240,3 +242,98 @@ def validate_parameter(self, parameter):
+        l_value = [index for (index, item) in enumerate(n_value) if item == 1]  # Locations of '1's
+
+        qc.x(qreg[l_value[1:k]])
+        m_current_value = 2 ** (l_value[0])
```
These could be removed if you want (same below)
```suggestion
        m_current_value = 2 ** l_value[0]
```

Cryoris commented on releasenotes/notes/uniform-superposition-gate-3bd95ffdf05ef18c.yaml:
```diff
@@ -0,0 +1,25 @@
+
+    .. code-block:: python
+
+       from qiskit import UniformSuperpositionGate, QuantumCircuit
```
`UniformSuperpositionGate` is not available as top level import, could you ensure this snippet runs as expected?

Hirmay replied:
Okay, I have done that.

Cryoris commented on test/python/circuit/test_uniform_superposition_gate.py:
```diff
@@ -0,0 +1,139 @@
+        qc = QuantumCircuit(n)
+        qc.append(gate, list(range(n)))
+        unitary_matrix = np.array(Operator(qc).data)
+        actual_sv = unitary_matrix[:, 0]
```
You should be able to just use
```suggestion
        actual_sv = Statevector(gate)
```
where `Statevector` lives in `qiskit.quantum_info`

Hirmay replied:
Okay, I have used that. Thank you, Julien!

Cryoris commented on test/python/circuit/test_uniform_superposition_gate.py:
```diff
@@ -0,0 +1,139 @@
+        np.testing.assert_allclose(desired_sv, actual_sv)
+
+    @data(2, 3, 5, 13)
+    def test_none_qubit_args(self, num_superpos_states):
```
I don't think we need to check that `func(None)` is the same as `func()` if the default arg is set to `None` 🙂 You can just remove this test.

Cryoris commented on test/python/circuit/test_uniform_superposition_gate.py:
```diff
@@ -0,0 +1,139 @@
+        np.testing.assert_allclose(desired_sv, actual_sv)
+
+    @data(2, 3, 5, 13)
+    def test_uniform_superposition_gate_transpile(self, num_superpos_states):
```
We can also remove this test as you only use standard Qiskit gates and the transpilation is already thoroughly tested by itself.

Hirmay replied:
Okay, I have removed this test. Thanks!

Cryoris reviewed (commented):
Thanks for taking care of the comments! There's some more steps we can take to clean up the code, then it looks good to go from my side 🙂

Hirmay committed:
Update state_preparation.py

Hirmay committed:
Update test_uniform_superposition_gate.py

Hirmay committed:
Update uniform-superposition-gate-3bd95ffdf05ef18c.yaml

Hirmay committed:
Update qiskit/circuit/library/data_preparation/state_preparation.py

Co-authored-by: Julien Gacon <gaconju@gmail.com>

Hirmay commented:
Hi @Cryoris, I have incorporated all the changes you suggested, and it will pass all the tests now. The inverse function in the latest release of Qiskit, caused one test to fail, but have fixed it now. Can you please look at it and see if there's anything else required?

Hirmay committed:
blacked state_preparation.py

Hirmay committed:
Update test_uniform_superposition_gate.py

Hirmay committed:
pylinted state_preparation.py

Hirmay committed:
removed transpile test_uniform_superposition_gate.py

Cryoris reviewed (approved):
Thanks for all the work and incorporating all the changes! This looks great 🙂

Hirmay commented:
Thank you so much to you, Jake, Luciano, and especially Steve, for going through my amateurishly written code and helping bring it to this stage, despite your incredibly busy schedules. 

Steve, I'm incredibly grateful for the support and assistance you've provided throughout this journey. Without you, it wouldn't have been possible.

I'll probably bother you guys again with another pull request 😅, hopefully next time it'll be easier to deal with.

woodsp-ibm commented:
Hirmay, you are very welcome. Congratulations on your successful contribution to Qiskit.

Hirmay commented:
Thanks, again, Steve! XD
